### PR TITLE
[G7C1-13] NEWACCOUNT Functionality

### PR DESCRIPTION
**Pull request by de-mendes**
# [G7C1-13](https://bath-group7.atlassian.net/browse/G7C1-13)

N.B. The title of PR must also reference a [JIRA ticket number] in square brackets.

## Description

* Includes the NEWACCOUNT functionality and error handling.

* Converts the project into a Maven project for easier build and testing by decoupling the build process of Client and Server.

* Refactors the accounts array of a Customer for a hasmap. Hashmaps have lower time complexity Big O in key retrieval (constant) than an array search O(n)

* Fixes some typos on ExampleClient

## Definition of Done

This PR:

- [x] Delivers the functionality of the ticket it closes.
- [x] Is tested locally and includes unit tests using the [JUnit 5](https://junit.org/junit5/) (where applicable).
- [x] Follows [Java style guide](https://google.github.io/styleguide/javaguide.html).
- [x] Code is commented using the [Javadoc tool](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html) and any other applicable documentation e.g. on Confluence is updated.
- [x] Is rebased/cleaned up to be easy to review.

### CODE REVIEW

**Review by AishhD — CHANGES_REQUESTED** [review]
@de-mendes I am finding it really hard to follow your commits especially the first one which has changes across many files. Is it possible to give your commits unique commit messages that describe what is happening in each commit or unique descripitions? This will make it easier to understand the changes you have made. Plus, if the commits are not structured like this already it would also be good to introduce only one change per commit to make it easier to review and debug in the future should we need to.

Also, you noted in the description the introduction of Maven into the project. Can you please explain what this is and how we can use it? Feel free to add a link the docs or any articles, etc that you found useful. This will help ensure we all have a shared understanding.

**Comment by de-mendes** [comment]
Hi @AishhD as a matter of fact, it is a common practice to have a single commit. My idea was to squash the three into one. However the number of commits shouldn't interfere with the review. With regards to the Maven information or documentation, it is all in the new Readme.

**Comment by AishhD** [comment]
> 
> 
> Hi @AishhD as a matter of fact, it is a common practice to have a single commit. My idea was to squash the three into one. However the number of commits shouldn't interfere with the review. With regards to the Maven information or documentation, it is all in the new Readme.

Hey @de-mendes

I hadn't spotted the maven information in the ReadMe, thanks for pointing it out.

 I'm a bit confused by your reply, you have 3 commits which all have the same title and no description, so it is unclear what each commit is doing. As your are creating changes across so many files it makes sense to have it broken into smaller commits, it is just unclear what each of these do. 

From your commits and titles I am unable to tell which commit is introducing Maven, where the new functionality is added, where errors are being handled and where the refactoring is being done. It may be helpful to think about it from a commit history perspective, the history would be unclear as there is no clear narrative about the changes introduced. For instance, if in two weeks we find the refactoring has a caused a bug we want to be able to look at the history and know from the history that this commit did some refactoring and is likely the cause of the bug.

**Comment by AishhD** [comment]
@de-mendes It may also make sense to introduce Maven in its own PR as I just had a look at the jira ticket and it is not clear from that why we would introduce Maven here.
Personally I'm struggling to understand why all the server files have moved and I thought you were introducing a testing framework but Maven seem to be much more

**Comment by de-mendes** [comment]
Hi @AishhD ,

Apologies for the lack of clarity earlier, I used the smartphone app and it is tougher to write a long message. 

What I was trying to say is that I was going to squash the commits into a single one. I have just done it, so now you'll see there's only one commit and no confusion with regards to the messages. 

With regards to commit messages here are the [conventions](https://www.conventionalcommits.org/en/v1.0.0/) I am following. Taking that into account, if your repository is using [semantic versioning](https://semver.org/) (widely used in the industry) different commit messages (feat, fix, BREAKING CHANGE) in a single PR,  might mess up with your versioning, that's why the Pull Request only has one change. I know we're not using semantic versioning but I am trying to follow best conventions.

Thank you for the link, I am happy to squash the single commit again and push different commits if that is going to make your review easier. 😊 I personally dont look at commit messages to review code, I just pull the branch and focus on the code as commit messages are subjective and can be misleading. 

PD: Some extra resources for MAVEN.
[Multi Module Maven Projects](https://www.baeldung.com/maven-multi-module)
[Maven DOCS](https://maven.apache.org/what-is-maven.html)

**Comment by AishhD** [comment]
@de-mendes thanks for the information on semantic versioning and how it affects commits. Its something I haven't heard of before
